macrosAutoCats/src/main/scala/com/softwaremill/macwire/autocats/internals/CatsProvidersGraphContext.scala

```diff
@@ -0,0 +1,186 @@
+      }
+
+    def topologicalSort(): List[Provider] = log.withBlock("Stable topological sort") {
+      def go(provider: Provider, usedProviders: Set[Provider]): List[Provider] =
+        log.withBlock(s"Going deeper for type [${provider.resultType}]") {
+
+          provider.dependencies.flatten.foldl(List.empty[Provider]) {
+            case (_, None) => c.abort(c.enclosingPosition, "Missing dependency.")
+            case (r, Some(p)) if (usedProviders ++ r.toSet).contains(p) =>
+              log.withBlock(s"Already used provider for type [${p.resultType}]")(r)
+            case (r, Some(p)) =>
+              log.withResult { (r ::: go(p, (usedProviders ++ r.toSet) + p)) :+ p }(result =>
+                s"Built list for the following types [${result.map(_.resultType).mkString(", ")}]"
+              )
+          }
+        }
+
+        val result = providers.foldLeft(List.empty[Provider]) {
+          case (resultProviders, nextProvider) if resultProviders.contains(nextProvider) => resultProviders
+          case (resultProviders, nextProvider) => resultProviders ::: go(nextProvider, resultProviders.toSet)
+        } :+ root
+
+        verifyOrder(result)
+
+        result
+    }
+
+  }
+
+  def buildGraph(rawProviders: List[c.universe.Expr[Any]], rootType: c.Type): CatsProvidersGraph = {
+    val (providers, fms): (List[Provider], List[FactoryMethodTree]) = rawProviders
+      .partitionBifold { expr =>
+        val tree = expr.tree
+        val tpe = typeCheckUtil.typeCheckIfNeeded(tree)
+
+        FactoryMethodTree(tree)
+          .map(_.asRight[Provider])
+          .orElse(Resource.fromTree(tree, tpe).map(_.asLeft[FactoryMethodTree]))
+          .orElse(Effect.fromTree(tree, tpe).map(_.asLeft[FactoryMethodTree]))
+          .getOrElse(new Instance(tree).asLeft[FactoryMethodTree])
+      }
+      .bimap(_.toList, _.toList)
+
+    log(s"Providers: [${providers.mkString(", ")}]")
+    log(s"Factory methods: [${fms.mkString(", ")}]")
+
+    val initContext = BuilderContext(providers, fms)
+    val (resolvedCtx, rootProvider) = maybeResolveParamWithCreator(resolveFactoryMethods(initContext))(rootType)
+      .getOrElse(c.abort(c.enclosingPosition, s"Cannot find a value of type: [$rootType]"))
+
+    val inputProvidersOrder = rawProviders
+      .map(expr => {
+        val tree = expr.tree
+        val tpe = typeCheckUtil.typeCheckIfNeeded(tree)
+        if (FactoryMethod.isFactoryMethod(tree)) FactoryMethod.underlyingResultType(tree)
+        else if (Resource.isResource(tpe)) Resource.underlyingType(tpe)
+        else if (Effect.isEffect(tpe)) Effect.underlyingType(tpe)
+        else tpe
+      })
+      .map(tpe =>
+        resolvedCtx.providers.find(_.resultType <:< tpe).getOrElse(c.abort(c.enclosingPosition, "Internal error"))
+      )
+
+    log(s"Input providers order [${inputProvidersOrder.mkString(", ")}]")
+
+    new CatsProvidersGraph(inputProvidersOrder ++ resolvedCtx.providers.diff(inputProvidersOrder), rootProvider)
```

Comment:
I'm not sure what's the purpose here of the `inputProvidersOrder` and the other providers? Can't we just pass the providers `providers ++ fms` to the graph and sort them?

Reply:
This point is necessary to sort independent providers in the same way they were passed (I think it's a good initial approach to the problem, because it makes the result order predictable). 
I was thinking about two other approaches to this problem:
* implicit order of providers that needs to be preserved during the whole resolution process - it would be definitely harder to maintain
* additional sorting logic in `topologicalSort` - I think it would harm `CatsProvidersGraph` encaptulation

Reply:
I still don't understand why we iterate over `rawProviders` twice :)

Reply:
Ah did you ask about these `providers` and `fms` -https://github.com/softwaremill/macwire/pull/188/files/a78010f1dd0991bf326716caacf5ff14946fb3b6#diff-8923920dc067a35890d1e81851c52f2387bb9d3e173c036c9a714372d7146558R61 ?
The `fms` name is actually incorrect, because it's a list of `FactoryMethodTree` that needs to be resolved in `resolveFactoryMethods` method. We cannot use providers from `BuilderContext` directly, because `resolveFactoryMethods` does not preserve the input order, so I think we cannot avoid the second iteration in a simple way.

Reply:
Ah did you ask about these `providers` and `fms` -https://github.com/softwaremill/macwire/pull/188/files/a78010f1dd0991bf326716caacf5ff14946fb3b6#diff-8923920dc067a35890d1e81851c52f2387bb9d3e173c036c9a714372d7146558R61 ?
The `fms` name is actually incorrect, because it's a list of `FactoryMethodTree` that needs to be resolved in `resolveFactoryMethods` method. We cannot use providers from `BuilderContext` directly, because `resolveFactoryMethods` does not preserve the input order, so I think we cannot avoid the second iteration in a simple way.

I extracted a common part to a variable.